### PR TITLE
Using size to quickcheck

### PR DESCRIPTION
**Pull request by MirkoBonadei**
The goal of this pull request is to have "sized generators" so that we can start a `ForAll` evaluation generating small values and increase their size as the property continues to hold.
This is the first step to improve our shrinking performances.

Notes of interest:
- the `Generator\int` and the `Generator\float` in other quickcheck's implementations are not generators based on a range , so now we are alligned with the others and if we want to create a generator based on a range we can use `Generator\choose` which is part of this pull request;
- `Generator\Natural` is not needed anymore, it's functionality is carried out by `Generator\Integer`

Some stuff must be improved before the merge:
- [x] the use of the size from the `ForAll` now grows linearly but it is not a big deal , we have to be more smart on that (exponential growth vs quadratic growth vs contigurable strategy);
- [x] fix `ShrinkingTimeLimitTest.php`, it is failing because now it takes more time to reach to counter-example. It should be fixed by the previous point or maybe it has to be tuned;
- [x] add configurable `maxSize` to `ForAll`, as default value 200 seems good;
- [x] rename `Sample::withSize` because now `size` term is overloaded;
- [x] `Sample::shrink` is ok, I can remove the todo because 10 is a good sample size to shrink from;
- [x] remove the skip flag to `examples/FloatTest.php`;
- [x] `Generator\Name` should exploit the generation size when invoked
- [x] review some TODOs
- [x] look some samples of generated values and improve the feature on the generators which are not using it at their best
- [ ] write some tests to accumulate metrics on the performance improvements (at which iteration we reach the counter-example? how long does it take to shrink it?)

### CODE REVIEW

**Comment by MirkoBonadei** [comment]
I have updated the todo list above after a live review.

**Comment by MirkoBonadei** [comment]
Here are some samples from the mainly used generators with size 10: https://gist.github.com/MirkoBonadei/b923dfb0ebe1ed120c25
And here some stuff with size 30: https://gist.github.com/MirkoBonadei/9c7e08c6b33c9d722484

Everything seems fine. The only thing to be clarified is that the `size` means "less than or equal to the size" for "scalar types" (eg: float, positive integer, negative integer, ecc...), while it means "of size" for composed types (eg: sequence, string).

We have also found a bug which I am already fixing, at the moment we consider 0 a good negative integer. :-)

**Comment by MirkoBonadei** [comment]
Samples of the two new generators `Generator\strictlyPos` and `Generator\strictlyNeg` with size 10: https://gist.github.com/MirkoBonadei/05f48f0a80cb3e7f199e

**Comment by MirkoBonadei** [comment]
The commit e1d8f08 fixes a possible issue with this statement:
_The only thing to be clarified is that the size means "less than or equal to the size" for "scalar types" (eg: float, positive integer, negative integer, ecc...), while it means "of size" for composed types (eg: sequence, string)_

But it is a fix only for `Generator\Sequence`. I think it should work also for `Generator\String` which now exposes Eris to the same risk explained in the e1d8f08's commit message, but probably there is something else to do with `Generator\String`, why it needs a `maximumLength`, what do you think if I drop it and the generation is only bounded between 0 and the generation size?

**Comment by giorgiosironi** [comment]
I agrees `Generator\String` should not need a `maximumLength`, now it is conceptually equivalent to a `Generator\Sequence` of `Generator\Char`. Just bounding in [0, $size] is fine

File: src/Eris/Quantifier/ForAll.php [hunk 1]
```diff
@@ -142,4 +155,57 @@ private function generatorsFrom($supposedToBeGenerators)
...
+        }
+        $this->sizes = $sizes;
+
+        return $this->sizes;
+    }
+
+    /**
+     * Returns the identity function.
+     */
+    private function lineraGrowth()
```
Comment (giorgiosironi) [inline src/Eris/Quantifier/ForAll.php]: typo in method name

**Comment by giorgiosironi** [comment]
:+1:

**Comment by MirkoBonadei** [comment]
I have just received an out of band :+1: by @gabrielelana :smile:, so we can merge.

**Pull request merged**